File: scripts/octave/methods/minimumSquarePolynomialRegressionMethod.m
function [result, err] = minimumSquarePolynomialRegressionMethod(x, y, polyDegree)
    err = ""; % Initialize err to an empty string

    [sampleSize] = length(x);

    % Checking if the given inputs has the same size
    if sampleSize != length(y)
        err = "x and y must have the same length";
        result = NaN;
        return;
    end

    n = polyDegree + 1;
    a = zeros(n, n);
    b = zeros(n, 1);

    % Generating the linear system
    b(1) = sum(y);
    a(1, 1) = sampleSize;
    for index = 1:n
        k1 = index - 1;
        k2 = index;
        for subIndex = 2:n
            sumAuxiliary = sum(power(x, k2));
            a(index, subIndex) = sumAuxiliary;
            a(subIndex, index) = a(index, subIndex);
            k2 += 1;
        end

        sumAuxiliary = 0;
        if index ~= 1
            for indexIterator = 1:1:sampleSize
                sumAuxiliary += y(indexIterator) * x(indexIterator)^(k1);
            end
            b(index) = sumAuxiliary;
        end
    end

    % Start the linear system solution

    % Superior Triangular Matrix
    [a, b] = solveSuperiorTriangle(n, a, b);

    % Start the system solution calculation
    result = zeros(n, 1);
    result(n) = b(n) / a(n, n);
    for k = n - 1:-1:1
        s = 0;
        for j = k + 1:n
            s = s + a(k, j) * result(j);
        end
        result(k) = (b(k) - s) / a(k, k);
    end
end

function [a, b] = solveSuperiorTriangle(n, a, b)
    for k = 1:n - 1
        for i = k + 1:n
            m = a(i, k) / a(k, k);
            a(i, k) = 0;
            for j = k + 1:n
                a(i, j) -= m * a(k, j);
            end
            b(i) = b(i) - m * b(k);
        end
    end
    return
end

function result = power(x, y)
    result = x.^y;
end

File: internal/methods/minimum_square_polynomial_regression_method.go
package methods

import (
	"errors"

	"github.com/jictyvoo/tec217_computing-methods/internal/models"
	"github.com/jictyvoo/tec217_computing-methods/internal/utils"
)

type MinimumSquarePolynomialRegressionMethod[T models.Numeric] struct {
	commonFuncZeroState[T]
}

func (mtd *MinimumSquarePolynomialRegressionMethod[T]) power(x, y T) T {
	var result T = 1
	for i := 0; i < int(y); i++ {
		result *= x
	}
	return result
}

func (mtd *MinimumSquarePolynomialRegressionMethod[T]) solveSuperiorTriangle(
	n int, a [][]T, b []T,
) {
	for k := 0; k < n-1; k++ {
		for i := k + 1; i < n; i++ {
			m := a[i][k] / a[k][k]

			// Change a matrix in-memory
			a[i][k] = 0
			for j := k + 1; j < n; j++ {
				a[i][j] -= m * a[k][j]
			}

			// Change b slice in-memory
			b[i] -= m * b[k]
		}
	}
}

func (mtd *MinimumSquarePolynomialRegressionMethod[T]) Run(
	x, y []T, polyDegree T,
) (result []T, err error) {
	sampleSize := len(x)
	// Check for consistent input
	if sampleSize != len(y) {
		err = errors.New("x and y must have the same length")
		return
	}

	n := int(polyDegree + 1)
	a := make([][]T, n)
	for i := range a {
		a[i] = make([]T, n)
	}
	a[0][0] = T(sampleSize)

	b := make([]T, n)

	// Generating a linear system
	for _, value := range y {
		b[0] += value
	}
	for index := 1; index <= n; index++ {
		k1 := T(index - 1)
		k2 := T(index)
		for subIndex := 2; subIndex <= n; subIndex++ {
			sumAuxiliary := utils.ReduceSlice(
				x, func(old, value T) T { return old + mtd.power(value, k2) },
			)
			a[index-1][subIndex-1] = sumAuxiliary
			a[subIndex-1][index-1] = a[index-1][subIndex-1]
			k2++
		}

		var sumAuxiliary T
		if index != 1 {
			for indexIterator := 0; indexIterator < sampleSize; indexIterator++ {
				sumAuxiliary += y[indexIterator] * mtd.power(x[indexIterator], k1)
			}
			b[index-1] = sumAuxiliary
		}
	}

	// Start the linear system solution

	// Superior Triangular Matrix
	mtd.solveSuperiorTriangle(n, a, b)

	// Start the system solution calculation
	result = make([]T, n)
	result[n-1] = b[n-1] / a[n-1][n-1]
	for k := n - 2; k >= 0; k-- {
		var s T
		for j := k + 1; j < n; j++ {
			s += a[k][j] * result[j]
		}
		result[k] = (b[k] - s) / a[k][k]
	}

	return
}
